### PR TITLE
CORE-1505: Support Multiple `invariant`s and Optional Labels

### PR DESCRIPTION
**Pull request by chrisnevers**

### CODE REVIEW

File: hs/src/Reach/AST/DL.hs [hunk 1]
```diff
@@ -104,7 +105,7 @@ data DLSStmt
   | DLS_While
       { dls_w_at :: SrcLoc
       , dls_w_asn :: DLAssignment
-      , dls_w_inv :: DLSBlock
+      , dls_w_inv :: [(DLSBlock, Maybe ByteString)]
```
Comment (jeapostrophe) [inline hs/src/Reach/AST/DL.hs:108]: You should make a type for this. A question I have now is whether any bindings are shared (I don't think they need to be, but I'm curious)

File: hs/src/Reach/Connector/ETH_Solidity.hs [hunk 2]
```diff
@@ -951,6 +951,8 @@ solCom = \case
   DL_Let _ (DLV_Let _ dv) (DLE_PrimOp _ GET_COMPANION []) -> do
     addMemVar dv
     solLargeArg dv $ mdaToMaybeLA T_Contract Nothing
+  DL_Let _ (DLV_Eff) (DLE_GetUntrackedFunds {}) ->
+    return ""
```
Comment (jeapostrophe) [inline hs/src/Reach/Connector/ETH_Solidity.hs:955]: Why this?

Comment (jeapostrophe) [reply hs/src/Reach/Connector/ETH_Solidity.hs:955]: An impossible? Did you add a test case?

File: hs/src/Reach/Eval/Core.hs [hunk 3]
```diff
@@ -5511,11 +5511,11 @@ doParallelReduce lhs (ParallelReduceRec {..}) = locAt slpr_at $ do
   let want lab = \case
         Just x -> return x
         Nothing -> expect_ $ Err_ParallelReduceIncomplete $ "missing " <> lab
-  inv_e <- want "invariant" pr_minv
```
Comment (jeapostrophe) [inline hs/src/Reach/Eval/Core.hs:5514]: The docs say it is allowed multiple times (which I read as 1+) but I think you might have implemented many (0+)

File: hs/t/n/Err_ParallelReduceIncomplete.txt [hunk 4]
```diff
@@ -1,4 +1,4 @@
-reachc: error[RE0083]: Parallel reduce incomplete: missing invariant
+reachc: error[RE0083]: Parallel reduce incomplete: missing while
```
Comment (jeapostrophe) [inline hs/t/n/Err_ParallelReduceIncomplete.txt:1]: This is still good I think

**Review by jeapostrophe — CHANGES_REQUESTED** [review]
You should look at examples and see if there are any cases where we have an && in the invariant and use this new feature

File: hs/src/Reach/Connector/ETH_Solidity.hs (hunk 2 above)
Comment (chrisnevers) [reply hs/src/Reach/Connector/ETH_Solidity.hs:955]: I discovered this while tweaking your duoswap PR to use `.api_`. When the shared checks and bindings are duplicated into PAY_EXPR, the result of `getUntrackedFunds` is unused and updated to `DLV_Eff`. Maybe, the purity of `getUntrackedFunds` should be handled specially during optimization so it doesn't make it this far?

File: hs/src/Reach/AST/DL.hs (hunk 1 above)
Comment (chrisnevers) [reply hs/src/Reach/AST/DL.hs:108]: OK & the bindings aren't shared

**Review by jeapostrophe — APPROVED** [review]

**Pull request merged**